classes/class-Admin.php

```diff
@@ -0,0 +1,807 @@
+		echo '</div></div></div>';
+
+		$tabindex = 5000;
+		foreach ( (array) $this->mt_seo_fields as $field_name => $field_data ) {
+			if ( empty( $global_values[ $field_name ] ) ) {
+				continue;
+			}
+
+			if ( 'textarea' === $field_data['input_type'] || 'text' === $field_data['input_type'] ) {
+				echo '<div class="form-field ' . esc_attr( $field_name ) . '"><h4><label for="' . esc_attr( $field_name ) . '">' . esc_html( $field_data['title'] ) . '</label></h4><div class="mt-form-field-contents"><p>';
+
+				if ( 'textarea' === $field_data['input_type'] ) {
+					echo '<textarea class="wide-seo-box" rows="4" cols="40" tabindex="' . esc_attr( $tabindex ) . '" name="' . esc_attr( $field_name ) . '" id="' . esc_attr( $field_name ) . '">' . esc_textarea( ${$field_name} ) . '</textarea>';
+				} elseif ( 'text' === $field_data['input_type'] ) {
+					echo '<input type="text" class="wide-seo-box" tabindex="' . esc_attr( $tabindex ) . '" name="' . esc_attr( $field_name ) . '" id="' . esc_attr( $field_name ) . '" value="' . esc_attr( ${$field_name} ) . '" />';
+				}
+				echo wp_kses(
+					'</p><p class="description">' . $field_data['desc'] . '</p></div></div>',
+					$this->form_get_kses_valid_tags__metabox_description()
+				);
+			}
+			$tabindex++;
+		}
+
+		wp_nonce_field( 'mt-seo', 'mt_seo_nonce', false );
+
+		/*
+		 * yoast functionality code is preserved from previous versions, updated only for CS. It is untested.
+		*/
+		// Remove old Yoast data
+		delete_post_meta( $post->ID, '_yoast_wpseo_metadesc' );
+		delete_post_meta( $post->ID, '_yoast_wpseo_title' );
+	}
+
+
+
+	/*******************************************************************************************************************
+	 * Methods related to saving options for the main options panel and the post meta box.
+	 */
+
+	/**
+	 * Calls the save routine if required. For backwards-compatibility reasons, each metatag's post meta value is
+	 * saved individually instead of as an array with the general options.
+	 *
+	 * @param int $post_id The post id the meta will be saved against.
+	 */
+	public function save_meta_for_singular_pages( $post_id ) {
+        // Bail if not a valid post type.
+		if ( ! isset( $_POST['post_type'] ) || ! $this->is_supported_post_type( $_POST['post_type'] ) ) { // @codingStandardsIgnoreLine: this is fine
+			return;
+		}
+
+		// Checks to make sure we came from the right page.
+		if ( ! wp_verify_nonce( $_POST['mt_seo_nonce'], 'mt-seo' ) ) { // @codingStandardsIgnoreLine: this is fine
+			return;
+		}
+
+		// If this is an autosave, our form has not been submitted so we don't want to do anything.
+		if ( defined( 'DOING_AUTOSAVE' ) && DOING_AUTOSAVE ) {
+		    return;
+		}
+
```

Comment:
Can we also add a check to ensure that the metadata isn't saved for the parent post on autosave?

Something like:

```suggestion
if ( wp_is_post_autosave( $post_id ) || wp_is_post_revision( $post_id ) ) {
    return;
}
```

Reply:
To confirm: by "parent post" you mean "the post containing the metata," not "The parent of the post being edited" correct? The suggested code seems to indicate the former. I'm curious why we would not want the metadata saved with an autosave or a revision?

Reply:
This is to ensure that the metadata doesn't save when the post is being previewed.

Reply:
ah, that is a good reason. Too bad there isn't a `wp_is_post_being_previewed()`

Reply:
Is `wp_is_post_autosave()` doing the same thing as `defined( 'DOING_AUTOSAVE' ) && DOING_AUTOSAVE ` immediately above?

Reply:
Doesn't seem to be: https://developer.wordpress.org/reference/functions/wp_is_post_autosave/

Reply:
done